Upgrade React Select

krishnaglick opened the pull request:
Resolves #16023

The old react-select setup made a lot of assumptions, abused certain facets of the library, and had some weird typing errors.
This upgrade is intended to put us in a better position for changes and improvements in the future. A number of the assumptions that were made still exist and will likely need a full rewrite in the future to increase maintainability.

For the future of our use of react-select, I feel like we should be able to get away from the whole `components={components}` and instead reference things more simply, and with better types. That would be a larger refactor that what's reasonable here however.

tealjulia commented on airbyte-webapp/src/components/base/DropDown/SelectContainer.tsx:
```diff
@@ -3,7 +3,14 @@ import { components, ContainerProps } from "react-select";
+export const SelectContainer: React.FC<
+  ContainerProps<OptionType, false> & {
+    selectProps: {
+      "data-testid": string;
```
Using an `aria-label` would be more accessible than a `data-testid` and should still be findable by React Testing Library and Cypress.

krishnaglick replied:
https://github.com/airbytehq/airbyte/blob/6b5e4fb6e6975d833f7c9054fc67e81ec1d53f8f/airbyte-webapp/src/views/Connector/ServiceForm/ServiceForm.test.tsx#L323
We're using this utility that doesn't offer an `aria-label` option. Changing this breaks the tests. Suggestions?

tealjulia replied:
We could query by role and use `{name: ${ariaLabelValue}}` probably.  However, this whole form appears to utilize `data-testid` widely so sounds like a bigger project for another time!

tealjulia commented:
A few notes:
- The issue mentioned nixing the styled components for dropdowns, but I still see `CustomSelect` is a styled component.
- Wow a bunch of the types got more complex... but seemingly more accurate as to what they're actually representing
- Wondering what you believe in here should be a future rewrite.  You mentioned it in the PR and I'm wondering what stands out to you here.

+ left one other specific comment about accessibility/testing

krishnaglick commented on airbyte-webapp/src/views/Connection/CatalogTree/components/SyncSettingsDropdown.tsx:
```diff
@@ -24,7 +24,8 @@ const Separator = styled.div`
 
-const SingleValue: React.FC<IProps> = (props) => {
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const SingleValue: React.FC<IProps<any>> = (props) => {
```
I think the weird component prop-drilling could be refactored out so these could be correctly typed.

krishnaglick replied:
https://github.com/airbytehq/airbyte/issues/16439

timroes replied:
While I agree with splitting this apart if possible, is there any way we can go with `unknown` for now. The `any` here currently actually allows us to access `props.data.value` below, while `props.data` could actually be undefined (at least according to types) and then throw us an exception at runtime.

krishnaglick replied:
Changed!

krishnaglick commented:
> A few notes:
> 
>     * The issue mentioned nixing the styled components for dropdowns, but I still see `CustomSelect` is a styled component.
> 
>     * Wondering what you believe in here should be a future rewrite.  You mentioned it in the PR and I'm wondering what stands out to you here.

1. Fixed, though it was less clean than I'd like.
2. I updated the PR description to include my thoughts.

tealjulia reviewed (changes requested):
Tested locally but haven't looked deeply at the code yet.

Looks like this dropdown on connector setup is missing a handful of styles:
![Screen Shot 2022-09-08 at 9 04 54 AM](https://user-images.githubusercontent.com/63751206/189129218-1e7a8dd7-6e9e-422e-8dc4-8b28384b1719.png)

krishnaglick commented:
> Tested locally but haven't looked deeply at the code yet.
> 
> Looks like at least this dropdown on connector setup is missing a handful of styles: ![Screen Shot 2022-09-08 at 9 04 54 AM](https://user-images.githubusercontent.com/63751206/189129218-1e7a8dd7-6e9e-422e-8dc4-8b28384b1719.png)

Pulled out the styled components removal work as it quickly got overly complex. This is "fixed"

tealjulia reviewed (approved):
Tested locally and did not find any visual or functional regression
